ocaml/testsuite/tests/typing-modular-explicit/general.ml

```diff
@@ -0,0 +1,797 @@
+Error: This expression has type 'a but an expression was expected of type A.t
+       The type constructor A.t would escape its scope
+|}]
+
+let f (x : (module T : Typ) -> _) : (module T : Typ) -> T.t = x
+
+[%%expect{|
+Line 1, characters 62-63:
+1 | let f (x : (module T : Typ) -> _) : (module T : Typ) -> T.t = x
+                                                                  ^
+Error: This expression has type (module T : Typ) -> 'a
+       but an expression was expected of type (module T : Typ) -> T.t
+       The module T would escape its scope
+|}]
+
+
+(* Testing the `S with type t = _` cases *)
+
+module type Coerce = sig
+  type a
+  type b
+  val coerce : a -> b
+end
+
+let coerce (module C : Coerce) x = C.coerce x
+
+module IntofBool = struct
+  type a = bool
+  type b = int
+  let coerce b = if b then 1 else 0
+end
+
+module BoolofInt = struct
+  type a = int
+  type b = bool
+  let coerce i = i <> 0
+end
+
+module IntofString = struct
+  type a = string
+  type b = int
+  let coerce = int_of_string
+end
+
+module IntofFloat = struct
+  type a = float
+  type b = int
+  let coerce = int_of_string_opt
+end
+
+[%%expect{|
+module type Coerce = sig type a type b val coerce : a -> b end
+val coerce : (module C : Coerce) -> C.a -> C.b = <fun>
+module IntofBool :
+  sig type a = bool type b = int val coerce : bool -> int end
+module BoolofInt :
+  sig type a = int type b = bool val coerce : int -> bool end
+module IntofString :
+  sig type a = string type b = int val coerce : string -> int end
+module IntofFloat :
+  sig type a = float type b = int val coerce : string -> int option end
+|}]
+
+let incr_general
+  (module Cfrom : Coerce with type b = int)
+  (module Cto : Coerce with type a = int and type b = Cfrom.a)
+  x =
+  coerce (module Cto) (1 + coerce (module Cfrom) x)
+
+[%%expect{|
+val incr_general :
+  (module Cfrom : Coerce with type b = int) ->
+  (module Cto : Coerce with type a = int and type b = Cfrom.a) ->
+  Cfrom.a -> Cto.b = <fun>
+|}]
+
+module type CoerceToInt = sig
+  type a
+  type b = int
+  val coerce : a -> int
+end
+
+module type CoerceFromInt = sig
+  type a = int
+  type b
+  val coerce : int -> b
+end
+
+let incr_general' :
+  (module C1 : CoerceToInt) -> (module C2 : CoerceFromInt with type b = C1.a) -> C1.a -> C1.a =
+  incr_general
+
+[%%expect{|
+module type CoerceToInt = sig type a type b = int val coerce : a -> int end
+module type CoerceFromInt = sig type a = int type b val coerce : int -> b end
+val incr_general' :
+  (module C1 : CoerceToInt) ->
+  (module C2 : CoerceFromInt with type b = C1.a) -> C1.a -> C1.a = <fun>
+|}]
+
+(* Recursive and mutually recursive definitions *)
+
+let rec f : (module T : Typ) -> int -> T.t -> T.t -> T.t =
+  fun (module T : Typ) n (x : T.t) (y : T.t) ->
+    if n = 0
+    then x
+    else f (module T) (n - 1) y x
+
+[%%expect{|
+val f : (module T : Typ) -> int -> T.t -> T.t -> T.t = <fun>
+|}]
+
+let rec f (module T : Typ) n (x : T.t) (y : T.t) =
```

Comment:
This would also impact the way polymorphic recursion is typed. Should we make this pass through another PR before hand to clearly distinguish both changes ?